dask_jobqueue/core.py

```diff
@@ -117,7 +122,6 @@ def __init__(self,
         self.worker_threads = threads
         self.name = name
 
-        self.jobs = dict()
```

Comment:
`self.jobs` was a mapping from n-->job_id. However, we were not really using it and it often was not cleaned up when a job ended (so I've removed it).

Reply:
I believe we can get rid of `self.n` and thus `self.jobs` as it is right now. However, for adaptive to work correctly, I believe we should keep track of all submitted jobs and their statuses. If not, don't we risk to continuously submit new jobs that are kept in the scheduler queue?

I'm in favour of keeping a dict mapping job_id --> job-status, e.g as @mrocklin proposed in #11: 'pending', 'running', 'finished' or equivalent. This way, in the scale_up method, we can take that into account.

A maybe simpler solution is to only keep track of the number of workers that are pending or running, and at  least to use this number in scale_up:
`return self.start_workers(n - number_of_pending_or_running_worker)`
But it seems difficult to deal with finished, running and pending worker this way.

Reply:
@guillaumeeb - I think we're in agreement here.

>  we should keep track of all submitted jobs and their statuses.

This would be nice but it might be somewhat difficult to do. We have three/four states that a job might be in:

- Pending - we may be able to combine some form of `qstat job_identifier` with a dictionary of submitted jobs `self.jobs above`
- Running - it is straight forward to determine which workers are attached to the scheduler
- Finished - jobs can exit normally or be killed (e.g. exceeded wall time). When JobQueue culls a worker, its easy to remove that worker from the list of jobs. However, when the queuing system kills a worker, we would need a way to remove that job from the list of running jobs. 

Generally, I think any use of `qstat` is going to be a bit ugly just because repeated queries of the queueing system tend to be quite expensive. For example:

```
$ time qstat 9595055
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
9595055.chadmin1  dask             jhamman           00:07:22 R premium

real	0m1.030s
user	0m0.012s
sys	0m0.000s
```

So we probably don't want to do this very often. Do others have ideas as to how one would track the status of these jobs in a tractable way?

Reply:
We could add a plugin to the scheduler that watched for when workers started and compared them against a set of known pending workers: http://distributed.readthedocs.io/en/latest/plugins.html

Something like the following:

```python
from distributed.diagnostics.plugin import SchedulerPlugin

class JobQueuePlugin(SchedulerPlugin):
    def add_worker(self, scheduler, worker=None, name=None, **kwargs):
        job_id = parse(name)
        pending_jobs.remove(job_id)

scheduler.add_plugin(JobQueuePlugin())
```

Reply:
This sounds like what we need. I can implement this as part of this PR.

Reply:
See 92eaf4e for an initial (untested) implementation using the scheduler plugin approach.